### PR TITLE
Add serial connector interface

### PR DESCRIPTION
**Pull request by ddebrunner**

### CODE REVIEW

File: connectors/serial/src/main/java/quarks/connectors/serial/SerialDevice.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+	 * Set the initialization function for this port.
+	 * Can be used to send setup instructions to the
+	 * device connected to this serial port.
+	 * <BR>
+	 * {@code initializer.accept(out, in)} is called once, passing the output
+	 * and input streams for this port.
+	 * 
+	 * @param initializer Function to be called when the application runs.
+	 */
+	public void setInitializer(BiConsumer<OutputStream,InputStream> initializer);
```
Comment (dlaboss) [inline connectors/serial/src/main/java/quarks/connectors/serial/SerialDevice.java]: It's early monday so I may just be a bit groggy...
I'm not understanding why it's desirable to have OutputStream and InputStream show up in this interface.  i.e., why isn't that info just a private aspect of the supplied `initializer` and `driver`?  `Runnable initializer`, `Supplier<T> driver`?
Also I imagine that it's common to have interrupt driven devices.  That would seem to imply an eventual need for a `pubic <T> TStream<T> events(...)`.  Maybe there will also be an eventual desire for a `generate(Supplier<T>)` like method?  At a high level, I'm missing the value of ~duplicating one or more stream-source methods present in `Topology`.

**Comment by ddebrunner** [comment]
Serial is more a means to an end, a serial port by itself doesn't do anything.

Streams are created by a protocol layer that uses the serial port, e.g. ELM327

image

So the ELM327 logic needs access to the OutputStream and InputStreams to send the commands to the ELM327 chip and receive the response.

It seemed to me that separating the protocol from connectivity to the serial port is a good thing, rather than having them intertwined.

**Comment by dlaboss** [comment]
I see.  I get the desire to separate opening/managing the {I,O} streams from a higher level device specific protocol layer.  

Independent of that it feels that it would be better if SerialDevice wasn't required to have ~copies of the Topology stream-source methods.  No? I also wonder if the signature of the `initializer` and `driver` functions isn't extensible enough.  Will the two streams alone always be enough?

Imagine something like the following (thinking out loud)...

```
interface SerialDevice {
    OutputStream getOutputStream();
    InputStream getInputStream();
    // maybe other things later?
    void setInitializer(Runnable initializer);
    <T> Supplier<T> wrap(Supplier<T> driver);  // returned supplier doesn't call driver until after initializer is run
}

Topology t = ...;
SerialDevice sd = ...;
MyProtocol myProtocol = new MyProtocol(sd);  // calls sd.setInitializer()
TStream<JsonObject> s = t.poll(t, myProtocol.getSupplier(), 200, MILLISECONDS);
// where myProtocol.getSupplier() {
//              return sd.wrap( () -> { use sd.get{*}Stream() and create and return the tuple } )};
```

**Comment by ddebrunner** [comment]
A serial port only has an input and an output I/O stream.

Like IotDevice, SerialDevice is a graph declare time construct, therefore it doesn't have InputStream and OutputStream instances, those are created at runtime. That's why I was passing them into the methods.

I'll think if it can be done some other way.

**Comment by ddebrunner** [comment]
Changed to a function model.

Added WIP as I've only changed the interface at this point, need to verify it will work with the implementation.

**Comment by ddebrunner** [comment]
ported over an implementation to the new style with no issue.

**Pull request merged**